Modify and fix new assumption facts

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #21228

#### Brief description of what is fixed or changed

1. New assumption keys for both old and new assumption system

`.is_positive_infinite` and `.is_negative_infinite` attributes are introduced. `Q.positive_infinite` and `Q.negative_infinite` keys are introduced.

2. New assumptions facts reorganization

Facts are now built in more systematical way, based on primitive predicates such as `negative_infinite`, `negative`, `zero`, `positive`, `positive_infinite`, and `imagniary`.

3. Bug fix

Incorrect result of `Q.infinite` is fixed. 

Before this change:
```python
>>> print(ask(Q.infinite(oo)))
None
```

After this change:
```python
>>> ask(Q.infinite(oo))
True
```

Also, incorrect results of `Q.hermitian` and `Q.antihermitian` are fixed. See #21228.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * `Q.positive_infinite` and `Q.negative_infinite` predicates are introduced.
  * `Q.infinite` now correctly evaluates to `True` for `oo`, `-oo`, and `zoo`.
  * `Q.hermitian(x)` now does not imply that `x` is antihermitian, and vice versa. `0` is now both hermitian and antihermitian.
 
* core
  * `.is_positive_infinite` and `.is_negative_infinite` attributes are introduced.
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
There was some discussion about this before. The difficulty is that the complexity of satask can grow exponentially in the number of predicates.

We need a way to translate these predicates into a smaller number before sending them into satask. I think that the way to do it is to have 5 primitive predicates:
```
neginf, neg, zero, pos, posinf
```
Then we have e.g.:
```
extended_real = neginf | neg | zero | pos | posinf
nonnegative = zero | positive
...
```

JSS95 commented:
Sounds good. There are a lot of discrepancies and awful errors in old and new assumption system. I think the facts in these two systems need thorough reorganization. At least, every key in old assumption system needs its counterpart in new assumption system.

JSS95 committed:
feat(assumptions) : add more old assumps to Q

Implement is_positive_infinite and is_negative_infinite
Implement Q.positive_infinite and Q.negative_infinite
Fix Q.infinite
Make assumptions.sathandlers._old_assump_replacer() simpler

JSS95 committed:
feat(assumptions) : add more old assumps to Q

Fix Q.hermitian and Q.antihermitian

JSS95 commented:
@oscarbenjamin 
`negative_infinite` and `positive_infinite` assumptions are introduced. Known facts are reorganized based on the primitive predicates. Plus, `Q.infinite`, `Q.hermitian` and `Q.antihermitian` predicates are fixed.

I sucessfully implemented `Q.extended_positive` and other extended real predicates, but that made the diff too large so I will open another PR for that after this one.

oscarbenjamin commented:
> Known facts are reorganized based on the primitive predicates.

Does this mean that they all get passed into satask though?

JSS95 committed:
feat(assumptions) : add more old assumps to Q

Make predicates decomposed into primitive predicates

This makes the facts used by satask simpler. Since the factbase of satask
is simplified, some queries cannot be directly computed by satask() and
thus tests for these queries are removed. However, these queries are
still computed by ask().

oscarbenjamin commented on sympy/core/assumptions.py:
```diff
@@ -257,6 +258,9 @@
     'extended_nonzero == extended_real & !zero',
+
+    'positive_infinite -> infinite',
+    'negative_infinite -> infinite',
```
I guess there could also be `positive_infinite -> extended_positive`

oscarbenjamin commented on sympy/assumptions/tests/test_satask.py:
```diff
@@ -143,15 +143,9 @@ def test_rational_irrational():
         Q.rational(z)) is None
-    assert satask(Q.irrational(x*y*z), Q.irrational(x) & Q.rational(y) &
-        Q.rational(z)) is True
-    assert satask(Q.irrational(pi*x*y), Q.rational(x) & Q.rational(y)) is True
```
Why are these being removed?

JSS95 replied:
Since the factbase of satask is simplified, some queries cannot be directly computed by `satask()` and
thus tests for these queries are removed. However, these queries are still computed by `ask()`.

oscarbenjamin replied:
Maybe I don't understand what it is that you have changed. I thought that the idea would just be for `satask` to convert incoming predicates like `Q.nonnegative(x)` into primitives like `Q.positive(x) | Q.zero(x)`.

Why has satask's handling of rational/irrational changed?

Are these tests also being tested for `ask` somewhere?

JSS95 replied:
My bad! This was not because of reduced factbase. It was because facts from sathandlers were not converted to primitive predicates. It's fixed now.

oscarbenjamin commented:
Changing anything in the core assumptions can impact performance. Can you run benchmarks comparing performance of this PR with master?

JSS95 committed:
feat(assumptions) : add more old assumps to Q

Add more facts to core/assumptions

JSS95 committed:
feat(assumptions) : add more old assumps to Q

Make pure predicates converted to primitives by to_NNF when possible

This makes tests from test_satask re-enabled.

JSS95 commented:
@oscarbenjamin 

This change slowed down the performance. I did benchmarking by running tests for `test_query.py` and `test_satask` several times over.

In master, `test_query` took about 10 sec and `test_satask` took about 0.85 sec.  
In this branch, `test_query` took about 16 sec and `test_satask` took about 1.2 sec.  

I think speed can be improved by letting `Q.real` and other excluded predicates into `get_known_facts_dict()`. This dict allows simple inference without using satask (please see the comment lines in `ask()`) so having more items here may show better result.

oscarbenjamin commented:
Have you tried running the benchmarks:
https://github.com/sympy/sympy_benchmarks

JSS95 committed:
Merge branch 'master' into oldassump_newassump

JSS95 commented:
Here is the highlighted benchmark compare result between master and this branch.

```

All benchmarks:

       before           after         ratio
     [85ff38c1]       [f89e0d20]
     <master>         <oldassump_newassump>

+        74.8±1ms       83.5±0.5ms     1.12  logic.LogicSuite.time_dpll2
+     3.49±0.02μs      3.84±0.03μs     1.10  matrices.TimeMatrixExpression.time_Mul
-      69.3±0.2ms       59.0±0.5ms     0.85  polygon.PolygonEnclosesPoint.time_bench01

```

oscarbenjamin commented:
I was asking more about the impact that changing the core assumptions would have from adding `is_positive/negative_infinite`. Both assumptions systems have the potential to slow down from the addition of new predicates.

JSS95 commented:
I will split down the PR and run benchmarks separately. This one is attempting to make too many things to benchmark each change efficiently.
